pkg/admission/server_test.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/lithammer/dedent"
+	"github.com/stretchr/testify/assert"
+	admission "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var decoder = codecs.UniversalDeserializer()
+
+func TestServeHTTPBasic(t *testing.T) {
+	assert := assert.New(t)
+	res := httptest.NewRecorder()
+	handler := http.HandlerFunc(ServeHTTP)
+	req, err := http.NewRequest("POST", "", nil)
+	req = req.WithContext(context.Background())
+	assert.Nil(err)
+	handler.ServeHTTP(res, req)
+	assert.Equal(400, res.Code)
+	assert.Equal("admission review object is missing\n",
+		res.Body.String())
+}
+
+func TestServeHTTPSubmissions(t *testing.T) {
+	for _, apiVersion := range []string{
+		"admission.k8s.io/v1beta1",
+		"admission.k8s.io/v1",
+	} {
+		for _, tt := range []struct {
+			name    string
+			reqBody string
+
+			wantRespCode        int
+			wantSuccessResponse admission.AdmissionResponse
+			wantFailureMessage  string
+		}{
```

Comment:
Nit: I think these tests might be easier to understand if these lists were assigned to variables and then looped through separately at the end.